cpp/src/strings/combine/concatenate_list_elements.cu

```diff
@@ -0,0 +1,266 @@
+  CUDF_EXPECTS(separator.is_valid(), "Parameter separator must be a valid string_scalar");
+
+  auto const num_rows = lists_strings_column.size();
+  if (num_rows == 0) { return detail::make_empty_strings_column(stream, mr); }
+
+  // Accessing the child strings column of the lists column must be done by calling `child()` on the
+  // lists column, not `get_sliced_child()`. This is because calling to `offsets_begin()` on the
+  // lists column returns a pointer to the offsets of the original lists column, which may not start
+  // from `0`.
+  auto const strings_col     = strings_column_view(lists_strings_column.child());
+  auto const lists_dv_ptr    = column_device_view::create(lists_strings_column.parent(), stream);
+  auto const strings_dv_ptr  = column_device_view::create(strings_col.parent(), stream);
+  auto const string_narep_dv = get_scalar_device_view(const_cast<string_scalar&>(narep));
+  auto const fn              = concat_strings_fn{*lists_dv_ptr,
+                                    lists_strings_column.offsets_begin(),
+                                    *strings_dv_ptr,
+                                    string_narep_dv,
+                                    string_view{separator.data(), separator.size()}};
+  auto [offsets_column, chars_column, null_mask, null_count] =
+    make_strings_children_with_null_mask(fn, num_rows, stream, mr);
+
+  return make_strings_column(num_rows,
+                             std::move(offsets_column),
+                             std::move(chars_column),
+                             null_count,
+                             std::move(null_mask),
+                             stream,
+                             mr);
+}
+
+namespace {
+/**
+ * @brief Concatenate strings functor using multiple separators.
+ *
+ * This will concatenate the strings within each list element of the given lists column
+ * and apply the separators. A unique separator is provided for each list row. The null-replacement
+ * string scalar `sep_narep_dv` (if valid) can be used in place of any null separator. The
+ * null-replacement string scalar `string_narep_dv` (if valid) is used in place of any string in a
+ * list that contains a null entry.
+ */
+struct concat_strings_multi_separators_fn {
+  column_device_view const lists_dv;
+  offset_type const* const list_offsets;
+  column_device_view const strings_dv;
+  column_device_view const separators_dv;
+  string_scalar_device_view const string_narep_dv;
+  string_scalar_device_view const sep_narep_dv;
+
+  offset_type* d_offsets{nullptr};
+
+  // If `d_chars == nullptr`: only compute sizes of the output strings.
+  // If `d_chars != nullptr`: only concatenate strings.
+  char* d_chars{nullptr};
+
+  __device__ void operator()(size_type idx)
+  {
+    if (lists_dv.is_null(idx) || (separators_dv.is_null(idx) && !sep_narep_dv.is_valid())) {
+      if (!d_chars) { d_offsets[idx] = size_type{-1}; }  // negative size means null string
```

Comment:
Using negative values here to identify nulls seems hacky since the offset value will need to be converted back to zero in the end.